File: src/main/scala/resonantinduction/mechanical/process/edit/TilePlacer.java
package resonantinduction.mechanical.process.edit;

import java.util.EnumSet;

import com.sun.org.apache.bcel.internal.generic.FieldGenOrMethodGen;
import cpw.mods.fml.common.network.ByteBufUtils;
import io.netty.buffer.ByteBuf;
import net.minecraft.block.Block;
import net.minecraft.block.material.Material;
import net.minecraft.client.renderer.texture.IIconRegister;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.init.Blocks;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.util.ChatComponentText;
import net.minecraft.util.IIcon;
import net.minecraft.world.IBlockAccess;
import net.minecraftforge.common.util.ForgeDirection;

import org.lwjgl.opengl.GL11;

import resonant.api.IRotatable;
import resonant.lib.network.discriminator.PacketTile;
import resonant.lib.network.discriminator.PacketType;
import resonant.lib.network.handle.IPacketReceiver;
import resonant.lib.render.RenderItemOverlayUtility;
import resonant.lib.utility.LanguageUtility;
import resonant.lib.utility.inventory.InternalInventoryHandler;
import resonant.lib.utility.inventory.InventoryUtility;
import universalelectricity.core.transform.vector.Vector3;

import cpw.mods.fml.relauncher.Side;
import cpw.mods.fml.relauncher.SideOnly;
import resonant.lib.content.prefab.java.TileInventory;

/**
 * @author tgame14
 * @since 18/03/14
 */
public class TilePlacer extends TileInventory implements IRotatable, IPacketReceiver
{
	@SideOnly(Side.CLIENT)
	private static IIcon iconFront, iconBack;
	private boolean doWork = false;
	private boolean autoPullItems = false;
	private byte placeDelay = 0;
	private InternalInventoryHandler invHandler;

	public TilePlacer()
	{
		super(Material.rock);
		setSizeInventory(1);
        normalRender(false);
        renderStaticBlock_$eq(true);
        this.rotationMask_$eq(Byte.parseByte("111111", 2));
	}

	public InternalInventoryHandler getInvHandler()
	{
		if (invHandler == null)
		{
			invHandler = new InternalInventoryHandler(this);
		}
		return invHandler;
	}

	@Override
	public void onAdded()
	{
		work();
	}

	@Override
	public void onNeighborChanged(Block block)
	{
		work();
	}

	@Override
	public void start()
	{
		super.start();
	}

	@Override
	public void update()
	{
        super.update();
		if (autoPullItems && this.ticks() % 5 == 0)
		{
			if (getStackInSlot(0) == null)
			{
				this.setInventorySlotContents(0, this.getInvHandler().tryGrabFromPosition(this.getDirection().getOpposite(), 1));
			}
		}
		if (doWork)
		{
			if (placeDelay < Byte.MAX_VALUE)
			{
				placeDelay++;
			}

			if (placeDelay >= 5)
			{
				doWork();
				doWork = false;
			}
		}

	}

	public void work()
	{
		if (isIndirectlyPowered())
		{
			doWork = true;
			placeDelay = 0;
		}
	}

	public void doWork()
	{
		//Tries to place the item stack into the world
		int side = 0;
		Vector3 placePos = position().add(getDirection());
		ItemStack placeStack = getStackInSlot(0);

		if (InventoryUtility.placeItemBlock(world(), placePos.xi(), placePos.yi(), placePos.zi(), placeStack, side))
		{
			if (placeStack.stackSize <= 0)
			{
				setInventorySlotContents(0, null);
			}

			markUpdate();
			doWork = false;
		}
	}

	@Override
	public boolean use(EntityPlayer player, int hitSide, Vector3 hit)
	{
		interactCurrentItem(this, 0, player);
		return true;
	}

	public boolean configure(EntityPlayer player, int side, Vector3 hit)
	{
		if (player.isSneaking())
		{
			this.autoPullItems = !this.autoPullItems;
			player.addChatComponentMessage(new ChatComponentText("AutoExtract: " + this.autoPullItems));
			return true;
		}
		return super.configure(player, side, hit);
	}

	@Override
	public PacketTile getDescPacket()
	{
		NBTTagCompound nbt = new NBTTagCompound();
		writeToNBT(nbt);
		return new PacketTile(this, nbt);
	}

    @Override
    public void onInventoryChanged()
    {
        sendPacket(getDescPacket());
    }

	@Override
	public void read(ByteBuf data, EntityPlayer player, PacketType type)
	{
		try
		{
			readFromNBT(ByteBufUtils.readTag(data));
		}
		catch (Exception e)
		{
			e.printStackTrace();
		}
	}

	@Override
	public void readFromNBT(NBTTagCompound nbt)
	{
		super.readFromNBT(nbt);
		this.autoPullItems = nbt.getBoolean("autoPull");
	}

	/**
	 * Writes a tile entity to NBT.
	 */
	@Override
	public void writeToNBT(NBTTagCompound nbt)
	{
		super.writeToNBT(nbt);
		nbt.setBoolean("autoPull", this.autoPullItems);
	}

	@Override
	public boolean canStore(ItemStack stack, int slot, ForgeDirection side)
	{
		return side == this.getDirection().getOpposite() && slot == 0;
	}

	@Override
	@SideOnly(Side.CLIENT)
	public IIcon getIcon(IBlockAccess access, int side)
	{
		int meta = access.getBlockMetadata(x(), y(), z());

		if (side == meta)
		{
			return iconFront;
		}
		else if (side == (meta ^ 1))
		{
			return iconBack;
		}

		return getIcon();
	}

	@Override
	@SideOnly(Side.CLIENT)
	public IIcon getIcon(int side, int meta)
	{
		if (side == (meta ^ 1))
		{
			return iconFront;
		}
		else if (side == meta)
		{
			return iconBack;
		}

		return getIcon();
	}

	@Override
	@SideOnly(Side.CLIENT)
	public void registerIcons(IIconRegister iconRegister)
	{
		super.registerIcons(iconRegister);
		iconFront = iconRegister.registerIcon(getTextureName() + "_front");
		iconBack = iconRegister.registerIcon(getTextureName() + "_back");
	}


    @Override
    public void renderDynamic(Vector3 position, float frame, int pass)
    {
        if (world() != null)
        {
            EnumSet set = EnumSet.allOf(ForgeDirection.class);
            set.remove(getDirection());
            set.remove(getDirection().getOpposite());
            set.remove(ForgeDirection.UP);
            set.remove(ForgeDirection.DOWN);
            GL11.glPushMatrix();
            RenderItemOverlayUtility.renderItemOnSides(this,  getStackInSlot(0), position.x(), position.y(), position.z(), LanguageUtility.getLocal("tooltip.noOutput"), set);
            GL11.glPopMatrix();
        }
    }
}

File: src/main/scala/edx/mechanical/machine/edit/TilePlacer.scala
package edx.mechanical.machine.edit

import java.util.EnumSet

import cpw.mods.fml.relauncher.{Side, SideOnly}
import net.minecraft.block.Block
import net.minecraft.block.material.Material
import net.minecraft.client.renderer.texture.IIconRegister
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.item.ItemStack
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.util.{ChatComponentText, IIcon}
import net.minecraft.world.IBlockAccess
import net.minecraftforge.common.util.ForgeDirection
import org.lwjgl.opengl.GL11
import resonantengine.lib.content.prefab.TInventory
import resonantengine.lib.network.discriminator.PacketTile
import resonantengine.prefab.network.TPacketSender
import resonantengine.lib.prefab.tile.spatial.ResonantTile
import resonantengine.lib.prefab.tile.traits.TRotatable
import resonantengine.lib.render.RenderItemOverlayUtility
import resonantengine.lib.transform.vector.Vector3
import resonantengine.lib.utility.LanguageUtility
import resonantengine.lib.utility.inventory.{InternalInventoryHandler, InventoryUtility}

/**
 * @author tgame14
 * @since 18/03/14
 */
object TilePlacer
{
  @SideOnly(Side.CLIENT) private var iconFront: IIcon = null
  @SideOnly(Side.CLIENT) private var iconBack: IIcon = null
}

class TilePlacer extends ResonantTile(Material.rock) with TInventory with TRotatable with TPacketSender
{
  private var _doWork: Boolean = false
  private var autoPullItems: Boolean = false
  private var placeDelay: Int = 0
  private var invHandler: InternalInventoryHandler = null

  //Constructor
  normalRender = false
  forceItemToRenderAsBlock = true
  renderStaticBlock = true
  this.rotationMask = 63

  override def getSizeInventory = 1

  override def onAdded
  {
    work
  }

  def work
  {
    if (isIndirectlyPowered)
    {
      _doWork = true
      placeDelay = 0
    }
  }

  override def onNeighborChanged(block: Block)
  {
    work
  }

  override def start
  {
    super.start
  }

  override def update
  {
    super.update
    if (autoPullItems && this.ticks % 5 == 0)
    {
      if (getStackInSlot(0) == null)
      {
        this.setInventorySlotContents(0, this.getInvHandler.tryGrabFromPosition(this.getDirection.getOpposite, 1))
      }
    }
    if (_doWork)
    {
      if (placeDelay < java.lang.Byte.MAX_VALUE)
      {
        placeDelay += 1
      }
      if (placeDelay >= 5)
      {
        doWork
        _doWork = false
      }
    }
  }

  def getInvHandler: InternalInventoryHandler =
  {
    if (invHandler == null)
    {
      invHandler = new InternalInventoryHandler(this)
    }
    return invHandler
  }

  def doWork
  {
    val side: Int = 0
    val placePos: Vector3 = toVector3.add(getDirection)
    val placeStack: ItemStack = getStackInSlot(0)
    if (InventoryUtility.placeItemBlock(world, placePos.xi, placePos.yi, placePos.zi, placeStack, side))
    {
      if (placeStack.stackSize <= 0)
      {
        setInventorySlotContents(0, null)
      }
      markUpdate
      _doWork = false
    }
  }

  override def use(player: EntityPlayer, hitSide: Int, hit: Vector3): Boolean =
  {
    interactCurrentItem(this, 0, player)
    return true
  }

  override def configure(player: EntityPlayer, side: Int, hit: Vector3): Boolean =
  {
    if (player.isSneaking)
    {
      this.autoPullItems = !this.autoPullItems
      player.addChatComponentMessage(new ChatComponentText("AutoExtract: " + this.autoPullItems))
      return true
    }
    return super.configure(player, side, hit)
  }

  override def getDescPacket: PacketTile =
  {
    val nbt: NBTTagCompound = new NBTTagCompound
    writeToNBT(nbt)
    return new PacketTile(this, nbt)
  }

  /**
   * Writes a tile entity to NBT.
   */
  override def writeToNBT(nbt: NBTTagCompound)
  {
    super.writeToNBT(nbt)
    nbt.setBoolean("autoPull", this.autoPullItems)
  }

  override def onInventoryChanged
  {
    sendDescPacket()
  }

  override def readFromNBT(nbt: NBTTagCompound)
  {
    super.readFromNBT(nbt)
    this.autoPullItems = nbt.getBoolean("autoPull")
  }

  override def canStore(stack: ItemStack, slot: Int, side: ForgeDirection): Boolean =
  {
    return side == this.getDirection.getOpposite && slot == 0
  }

  @SideOnly(Side.CLIENT) override def getIcon(access: IBlockAccess, side: Int): IIcon =
  {
    val meta: Int = access.getBlockMetadata(xi, yi, zi)
    if (side == meta)
    {
      return TilePlacer.iconFront
    }
    else if (side == (meta ^ 1))
    {
      return TilePlacer.iconBack
    }
    return getIcon
  }

  @SideOnly(Side.CLIENT) override def getIcon(side: Int, meta: Int): IIcon =
  {
    if (side == (meta ^ 1))
    {
      return TilePlacer.iconFront
    }
    else if (side == meta)
    {
      return TilePlacer.iconBack
    }
    return getIcon
  }

  @SideOnly(Side.CLIENT) override def registerIcons(iconRegister: IIconRegister)
  {
    super.registerIcons(iconRegister)
    TilePlacer.iconFront = iconRegister.registerIcon(getTextureName + "_front")
    TilePlacer.iconBack = iconRegister.registerIcon(getTextureName + "_back")
  }

  override def renderDynamic(position: Vector3, frame: Float, pass: Int)
  {
    if (world != null)
    {
      val set: EnumSet[ForgeDirection] = EnumSet.allOf(classOf[ForgeDirection])
      set.remove(getDirection)
      set.remove(getDirection.getOpposite)
      set.remove(ForgeDirection.UP)
      set.remove(ForgeDirection.DOWN)
      GL11.glPushMatrix
      RenderItemOverlayUtility.renderItemOnSides(this, getStackInSlot(0), position.x, position.y, position.z, LanguageUtility.getLocal("tooltip.noOutput"), set)
      GL11.glPopMatrix
    }
  }
}